File: src/java/com/sapienter/jbilling/server/user/db/CreditCardDTO.java
/*
 * JBILLING CONFIDENTIAL
 * _____________________
 *
 * [2003] - [2012] Enterprise jBilling Software Ltd.
 * All Rights Reserved.
 *
 * NOTICE:  All information contained herein is, and remains
 * the property of Enterprise jBilling Software.
 * The intellectual and technical concepts contained
 * herein are proprietary to Enterprise jBilling Software
 * and are protected by trade secret or copyright law.
 * Dissemination of this information or reproduction of this material
 * is strictly forbidden.
 */
package com.sapienter.jbilling.server.user.db;

import java.io.Serializable;
import java.util.Date;
import java.util.HashSet;
import java.util.Set;

import javax.persistence.CascadeType;
import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.JoinTable;
import javax.persistence.ManyToMany;
import javax.persistence.OneToMany;
import javax.persistence.Table;
import javax.persistence.TableGenerator;
import javax.persistence.Transient;
import javax.persistence.Version;

import org.apache.log4j.Logger;
import com.sapienter.jbilling.common.FormatLogger;
import com.sapienter.jbilling.common.Constants;
import com.sapienter.jbilling.common.Util;
import com.sapienter.jbilling.server.payment.db.PaymentDTO;
import com.sapienter.jbilling.server.util.Context;
import com.sapienter.jbilling.server.util.DataEncrypter;

@Entity
@TableGenerator(
        name = "credit_card_GEN",
        table = "jbilling_seqs",
        pkColumnName = "name",
        valueColumnName = "next_id",
        pkColumnValue = "credit_card",
        allocationSize = 100)
@Table(name = "credit_card")
public class CreditCardDTO implements Serializable {

    private static final FormatLogger LOG = new FormatLogger(Logger.getLogger(CreditCardDTO.class));

    private static final String OBSCURED_NUMBER_FORMAT = "************"; // + last four digits

    private int id;
    private String ccNumber;
    private Date ccExpiry;
    private String name;
    private Integer ccType;
    private int deleted;
    private String securityCode;
    private String ccNumberPlain;
    private String gatewayKey;
    private Set<PaymentDTO> payments = new HashSet<PaymentDTO>(0);
    private Set<UserDTO> baseUsers = new HashSet<UserDTO>(0);
    private Integer versionNum;
    private boolean hasChanged= false;

    public CreditCardDTO() {
    }

    public CreditCardDTO(int id, String ccNumber, Date ccExpiry, int ccType, int deleted) {
        this.id = id;
        this.ccNumber = ccNumber;
        this.ccExpiry = ccExpiry;
        this.ccType = ccType;
        this.deleted = deleted;
    }

    public CreditCardDTO(int id, String ccNumber, Date ccExpiry, String name,
            int ccType, int deleted, String securityCode,
            String ccNumberPlain) {
        this.id = id;
        this.ccNumber = ccNumber;
        this.ccExpiry = ccExpiry;
        this.name = name;
        this.ccType = ccType;
        this.deleted = deleted;
        this.securityCode = securityCode;
        this.ccNumberPlain = ccNumberPlain;
    }

    public CreditCardDTO(com.sapienter.jbilling.server.entity.CreditCardDTO oldCC) {
        this.id = oldCC.getId() == null ? 0 : oldCC.getId();
        this.ccExpiry = oldCC.getExpiry();
        this.securityCode = oldCC.getSecurityCode();
        this.ccType = oldCC.getType();
        this.deleted = oldCC.getDeleted() == null ? 0 : oldCC.getDeleted();
        this.gatewayKey = oldCC.getGatewayKey();
        this.hasChanged = oldCC.getHasChanged();

        setName(oldCC.getName());
        setNumber(oldCC.getNumber());
    }

    @Id
    @GeneratedValue(strategy = GenerationType.TABLE, generator = "credit_card_GEN")
    @Column(name = "id", unique = true, nullable = false)
    public int getId() {
        return this.id;
    }

    public void setId(int id) {
        this.id = id;
    }

    /**
     * Returns the raw encrypted credit card number as persisted to the database.
     * Do not use unless you know what you are doing, use {@link #getNumber()} instead.
     *
     * @return raw encrypted credit card number
     */
    @Column(name = "cc_number", nullable = false, length = 100)
    public String getRawNumber() {
        return this.ccNumber;
    }

    protected void setRawNumber(String ccNumber) {
        this.ccNumber = ccNumber;
    }

    @Column(name = "cc_expiry", nullable = false, length = 13)
    public Date getCcExpiry() {
        return this.ccExpiry;
    }

    public void setCcExpiry(Date ccExpiry) {
        this.ccExpiry = ccExpiry;
    }

    @Transient
    public Date getExpiry() {
        return getCcExpiry();
    }

    public void setExpiry(Date ccExpiry) {
        setCcExpiry(ccExpiry);
    }

    @Transient
    public String getName() {
        if (getRawName() == null) return null;
        return ((DataEncrypter)Context.getBean(Context.Name.DATA_ENCRYPTER)).decrypt(getRawName());
    }

    public void setName(String name) {
        if (name == null) {
            setRawName(null);
            return;
        }
        setRawName(((DataEncrypter)Context.getBean(Context.Name.DATA_ENCRYPTER)).encrypt(name));
        // validate that the saved name is retrivable and valid
        try {
            if (!getName().equals(name)) {
                LOG.error("The credit card name " + name + " was wrongly encrypted to " + getName());
            }
        } catch (Exception e) {
            LOG.error("The credit card name " + name + " was wrongly encrypted to " + getName());
        }
    }

    @Column(name = "name", length = 150)
    protected String getRawName() {
        return this.name;
    }

    protected void setRawName(String name) {
        this.name = name;
    }

    @Transient
    public Integer getType() {
        return getCcType();
    }

    @Column(name = "cc_type", nullable = false)
    public Integer getCcType() {
        return this.ccType;
    }

    /**
     * Sets the credit card type. Note that the {@link #setNumber(String)} attempts to auto-detect
     * this value whenever a credit card number is set, eliminating the need to explicity
     * set the type.
     *
     * @see com.sapienter.jbilling.common.CommonConstants PAYMENT_METHOD*
     * @param ccType credit card type
     */
    public void setCcType(Integer ccType) {
        this.ccType = ccType;
    }

    @Column(name = "deleted", nullable = false)
    public int getDeleted() {
        return this.deleted;
    }

    public void setDeleted(int deleted) {
        this.deleted = deleted;
    }

    @Transient
    public String getSecurityCode() {
        return this.securityCode;
    }

    public void setSecurityCode(String securityCode) {
        this.securityCode = securityCode;
    }

    @Column(name = "cc_number_plain", length = 20)
    public String getCcNumberPlain() {
        return this.ccNumberPlain;
    }

    public void setCcNumberPlain(String ccNumberPlain) {
        this.ccNumberPlain = ccNumberPlain;
    }

    /**
     * Returns a vendor specific gateway key that can be used to authenticate payments
     * without sending the rest of the credit card details.
     *
     * @return unique gateway key
     */
    @Column(name = "gateway_key")
    public String getGatewayKey() {
        return gatewayKey;
    }

    public void setGatewayKey(String gatewayKey) {
        this.gatewayKey = gatewayKey;
    }

    /**
     * Returns true if this credit card should be handled using the stored gateway key instead
     * of the stored credit card number. This usually means that the credit card number has
     * been obscured and cannot be used to make a payment.
     *
     * @return true if gateway key should be used for payment, false if not
     */
    @Transient
    public boolean useGatewayKey() {        
        return (Constants.PAYMENT_METHOD_GATEWAY_KEY.equals(getCcType()) || getGatewayKey() != null);
    }

    @OneToMany(cascade = CascadeType.ALL, fetch = FetchType.LAZY, mappedBy = "creditCard")
    public Set<PaymentDTO> getPayments() {
        return this.payments;
    }

    public void setPayments(Set<PaymentDTO> payments) {
        this.payments = payments;
    }

    @ManyToMany(cascade = CascadeType.ALL, fetch = FetchType.LAZY)
    @JoinTable(name = "user_credit_card_map", joinColumns = {@JoinColumn(name = "credit_card_id", updatable = false)}, inverseJoinColumns = {@JoinColumn(name = "user_id", updatable = false)})
    public Set<UserDTO> getBaseUsers() {
        return this.baseUsers;
    }

    public void setBaseUsers(Set<UserDTO> baseUsers) {
        this.baseUsers = baseUsers;
    }

    @Version
    @Column(name = "OPTLOCK")
    public Integer getVersionNum() {
        return versionNum;
    }

    public void setVersionNum(Integer versionNum) {
        this.versionNum = versionNum;
    }

    @Transient
    public String getNumber() {
        if (getRawNumber() == null) return null;
        return ((DataEncrypter)Context.getBean(Context.Name.DATA_ENCRYPTER)).decrypt(getRawNumber());
    }

    /**
     * This method sets the current credit card number in an encrypted form,
     * and attempts to auto-detect the credit card type from the given number.
     *
     * @see Util#getPaymentMethod(String)
     * @param number credit card number
     */
    @Transient
    public void setNumber(String number) {
        if (number == null || number.trim().equals("")) {
            setRawNumber(null);
            setCcNumberPlain(null);
        } else {
            String crip = ((DataEncrypter)Context.getBean(Context.Name.DATA_ENCRYPTER)).encrypt(number);
            setRawNumber(crip);
            setCcNumberPlain(number.substring(number.length() - 4));
            setCcType(Util.getPaymentMethod(number));
        }
    }

    /**
     * Obscures this credit cards stored number by replacing all but the last four digits
     * with asterisks (*). This method will overwrite the current number rendering this
     * credit card invalid for anything other than display purposes. 
     */
    public void obscureNumber() {
        if (getRawNumber() != null && getCcNumberPlain() != null) {
            setNumber(OBSCURED_NUMBER_FORMAT + getCcNumberPlain());
        }
    }

    /**
     * Returns true if this credit card has been obscured (part of the number
     * has been masked with asterisks '*') false if not, or if the number is null.
     * @return true if masked, false if not or number is null.
     */
    @Transient
    public boolean isNumberObsucred() {
        return getRawNumber() != null && getNumber().contains("*");
    }

    @Transient
    public com.sapienter.jbilling.server.entity.CreditCardDTO getOldDTO() {
        com.sapienter.jbilling.server.entity.CreditCardDTO oldCC =
                new com.sapienter.jbilling.server.entity.CreditCardDTO();

        oldCC.setId(this.getId());
        oldCC.setDeleted(this.getDeleted());
        oldCC.setExpiry(this.getExpiry());
        oldCC.setName(this.getName());
        oldCC.setNumber(this.getNumber());
        oldCC.setSecurityCode(this.getSecurityCode());
        oldCC.setType(this.getCcType());
        oldCC.setGatewayKey(this.getGatewayKey());

        return oldCC;
    }

    @Transient
    public Boolean getHasChanged() {
        return hasChanged;
    }

    @Transient
    public void setHasChanged(Boolean hasChanged) {
        this.hasChanged = hasChanged;
    }
}

File: src/php/src/CreditCardDTO.php
<?php
/**
 * jbilling-php-api
 * Copyright (C) 2007-2009  Make A Byte, inc
 * http://www.makeabyte.com
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 * 
 * @package com.makeabyte.contrib.jbilling.php
 */

/**
  * CreditCardDTO
  * @author Jeremy Hahn
  * @version 1.0
  * @copyright Make A Byte, inc
  * @package com.makeabyte.contrib.jbilling.php
  */

class CreditCardDTO {

      var $deleted;          // If the record has been deleted, this field contains “1”, otherwise it contains “0”. Note that deletion cannot be carried out by simply setting a “1” in this field.
      var $expiry;           // Expiration date of the credit card. Usually, card expiration dates are expressed in month/year form, such as “05/11” or “May 2011”. This field contains the last day the card is valid, in this example, “05/31/2011”.
      var $id;               // Unique identifier for this record.
      var $name;             // Credit card owner's name. This is the name that appears physically on the credit card.
      var $number;           // Credit card number. Usually, a 16 digit number.
      var $securityCode;     // CCV (Credit Card Verification) code of the credit card.
      var $type;             // Credit Card type. See Appendix A for acceptable values.

      /**
       * The CreditCardDTO constructor
       * 
       * @access public
       */
	  public function CreditCardDTO() {
	  }
	  /**
	   * Sets the deleted property of the CreditCardDTO object
	   * 
	   * @access public
	   * @param Integer $int If the record has been deleted, this field contains “1”, otherwise it contains “0”. Note that deletion cannot be carried out by simply setting a “1” in this field.
	   * @return void
	   */
	  public function setDeleted( $int ) {
	  	
	  	     $this->deleted = $int;
	  }
	  /**
	   * Sets the expiry property of the CreditCardDTO object
	   * 
	   * @access public
	   * @param Date $date Expiration date of the credit card. Usually, card expiration dates are expressed in month/year form, such as “05/11” or “May 2011”. This field contains the last day the card is valid, in this example, “05/31/2011”.
	   * @return void
	   */
	  public function setExpiry( $date ) {
	  	
	  	     $this->expiry = $date;
	  }
	  /**
	   * Sets the id property of the CreditCardDTO object
	   * 
	   * @access public
	   * @param Integer $id Unique identifier for this record.
	   * @return void
	   */
	  public function setId( $id ) {
	  	
	  	     $this->id = $id;
	  }
	  /**
	   * Sets the name property of the CreditCardDTO object
	   * 
	   * @access public
	   * @param String $name Credit card owner's name. This is the name that appears physically on the credit card.
	   * @return void
	   */
	  public function setName( $name ) {
	  	
	  	     $this->name = $name;
	  }
	  /**
	   * Sets the number property of the CreditCardDTO object
	   * 
	   * @access public
	   * @param String $number Credit card number. Usually, a 16 digit number.
	   * @return void
	   */
	  public function setNumber( $number ) {
	  	
	  	     $this->number = $number;
	  }
	  /**
	   * Sets the securityCode property of the CreditCardDTO object
	   * 
	   * @access public
	   * @param Integer $code CCV (Credit Card Verification) code of the credit card.
	   * @return void
	   */
	  public function setSecurityCode( $code ) {
	  	
	  	     $this->securityCode = $code;
	  }
	  /**
	   * Sets the type property of the CreditCardDTO object
	   * 
	   * @access public
	   * @param Integer $type Credit Card type. See Appendix A for acceptable values.
	   * @return void
	   */
	  public function setType( $type ) {
	  	
	  	     $this->type = $type;
	  }
	  /**
	   * Gets the deleted property of the CreditCardDTO object
	   * 
	   * @access public
	   * @return Integer If the record has been deleted, this field contains “1”, otherwise it contains “0”. Note that deletion cannot be carried out by simply setting a “1” in this field.
	   */
	  public function getDeleted() {
	  	
	  	     return $this->deleted;
	  }
	  /**
	   * Gets the expiry property of the CreditCardDTO object
	   * 
	   * @access public
	   * @return Date Expiration date of the credit card. Usually, card expiration dates are expressed in month/year form, such as “05/11” or “May 2011”. This field contains the last day the card is valid, in this example, “05/31/2011”.
	   */
	  public function getExpiry() {
	  	
	  	     return $this->expiry;
	  }
	  /**
	   * Gets the id property of the CreditCardDTO object
	   * 
	   * @access public
	   * @return Integer Unique identifier for this record.
	   */
	  public function getId() {
	  	
	  	     return $this->id;
	  }
	  /**
	   * Gets the name property of the CreditCardDTO object
	   * 
	   * @access public
	   * @return String Credit card owner's name. This is the name that appears physically on the credit card.
	   */
	  public function getName() {
	  	
	  	     return $this->name;
	  }
	  /**
	   * Gets the number property of the CreditCardDTO object
	   * 
	   * @access public
	   * @return String Credit card number. Usually, a 16 digit number.
	   */
	  public function getNumber() {
	  	
	  	     return $this->number;
	  }
	  /**
	   * Gets the securityCode property of the CreditCardDTO object
	   * 
	   * @access public
	   * @return Integer CCV (Credit Card Verification) code of the credit card.
	   */
	  public function getSecurityCode() {
	  	
	  	     return $this->securityCode;
	  }
	  /**
	   * Gets the type property of the CreditCardDTO object
	   * 
	   * @access public
	   * @return Integer Credit Card type. See Appendix A for acceptable values.
	   */
	  public function getType() {
	  	
	  	     return $this->type;
	  }	    	  
}
?>